In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib as mpl
from scipy.stats import stats
import warnings
warnings.filterwarnings("ignore")
from scipy.stats.stats import pearsonr


In [2]:
df = pd.read_csv("MDG_Export_20191227 (1).csv")

In [3]:
df.head()

CountryCode      Country  SeriesCode MDG  \
0            4  Afghanistan         567   Y   
1            4  Afghanistan         749   Y   
2            4  Afghanistan         616   Y   
3            8      Albania         567   Y   
4            8      Albania         749   Y   

                                              Series    1990 Footnotes Type  \
0  Proportion of land area covered by forest, per...     2.1       NaN    E   
1  Carbon dioxide emissions (CO2), thousand metri...  2676.9       NaN    G   
2  Terrestrial and marine areas protected to tota...    0.36       NaN    C   
3  Proportion of land area covered by forest, per...    28.8       NaN    C   
4  Carbon dioxide emissions (CO2), thousand metri...    7488       NaN    G   

     1991 Footnotes.1  ... Type.26 2017 Footnotes.27 Type.27 2018  \
0                      ...                                          
1  2493.6         NaN  ...                                          
2                      ...                                          
3                      ...                                          
4  3971.4         NaN  ...                                          

  Footnotes.28 Type.28 2019 Footnotes.29 Type.29  
0                                                 
1                                                 
2                                                 
3                                                 
4                                                 

[5 rows x 95 columns]

In [4]:
df.tail()

CountryCode   Country  SeriesCode MDG  \
685          894    Zambia         749   Y   
686          894    Zambia         616   Y   
687          716  Zimbabwe         567   Y   
688          716  Zimbabwe         749   Y   
689          716  Zimbabwe         616   Y   

                                                Series     1990 Footnotes  \
685  Carbon dioxide emissions (CO2), thousand metri...   2445.9       NaN   
686  Terrestrial and marine areas protected to tota...    36.05       NaN   
687  Proportion of land area covered by forest, per...     57.3       NaN   
688  Carbon dioxide emissions (CO2), thousand metri...  15504.1       NaN   
689  Terrestrial and marine areas protected to tota...     16.9       NaN   

    Type     1991 Footnotes.1  ... Type.26 2017 Footnotes.27 Type.27 2018  \
685    G   2416.6         NaN  ...                                          
686    C                       ...                                          
687    C                       ...                                          
688    G  15823.1         NaN  ...                                          
689    C                       ...                                          

    Footnotes.28 Type.28 2019 Footnotes.29 Type.29  
685                                                 
686                                                 
687                                                 
688                                                 
689                                                 

[5 rows x 95 columns]

# Cleaning And Preparing Data

In [5]:
df1 = df.replace(r'^\s*$', np.nan, regex = True)
df1

CountryCode      Country  SeriesCode MDG  \
0              4  Afghanistan         567   Y   
1              4  Afghanistan         749   Y   
2              4  Afghanistan         616   Y   
3              8      Albania         567   Y   
4              8      Albania         749   Y   
..           ...          ...         ...  ..   
685          894       Zambia         749   Y   
686          894       Zambia         616   Y   
687          716     Zimbabwe         567   Y   
688          716     Zimbabwe         749   Y   
689          716     Zimbabwe         616   Y   

                                                Series     1990 Footnotes  \
0    Proportion of land area covered by forest, per...      2.1       NaN   
1    Carbon dioxide emissions (CO2), thousand metri...   2676.9       NaN   
2    Terrestrial and marine areas protected to tota...     0.36       NaN   
3    Proportion of land area covered by forest, per...     28.8       NaN   
4    Carbon dioxide emissions (CO2), thousand metri...     7488       NaN   
..                                                 ...      ...       ...   
685  Carbon dioxide emissions (CO2), thousand metri...   2445.9       NaN   
686  Terrestrial and marine areas protected to tota...    36.05       NaN   
687  Proportion of land area covered by forest, per...     57.3       NaN   
688  Carbon dioxide emissions (CO2), thousand metri...  15504.1       NaN   
689  Terrestrial and marine areas protected to tota...     16.9       NaN   

    Type     1991  Footnotes.1  ... Type.26 2017  Footnotes.27 Type.27 2018  \
0      E      NaN          NaN  ...     NaN  NaN           NaN     NaN  NaN   
1      G   2493.6          NaN  ...     NaN  NaN           NaN     NaN  NaN   
2      C      NaN          NaN  ...     NaN  NaN           NaN     NaN  NaN   
3      C      NaN          NaN  ...     NaN  NaN           NaN     NaN  NaN   
4      G   3971.4          NaN  ...     NaN  NaN           NaN     NaN  NaN   
..   ...      ...          ...  ...     ...  ...           ...     ...  ...   
685    G   2416.6          NaN  ...     NaN  NaN           NaN     NaN  NaN   
686    C      NaN          NaN  ...     NaN  NaN           NaN     NaN  NaN   
687    C      NaN          NaN  ...     NaN  NaN           NaN     NaN  NaN   
688    G  15823.1          NaN  ...     NaN  NaN           NaN     NaN  NaN   
689    C      NaN          NaN  ...     NaN  NaN           NaN     NaN  NaN   

     Footnotes.28 Type.28 2019  Footnotes.29 Type.29  
0             NaN     NaN  NaN           NaN     NaN  
1             NaN     NaN  NaN           NaN     NaN  
2             NaN     NaN  NaN           NaN     NaN  
3             NaN     NaN  NaN           NaN     NaN  
4             NaN     NaN  NaN           NaN     NaN  
..            ...     ...  ...           ...     ...  
685           NaN     NaN  NaN           NaN     NaN  
686           NaN     NaN  NaN           NaN     NaN  
687           NaN     NaN  NaN           NaN     NaN  
688           NaN     NaN  NaN           NaN     NaN  
689           NaN     NaN  NaN           NaN     NaN  

[690 rows x 95 columns]

In [6]:
#Removing unncessary columns

df1.drop(["Footnotes",
         "Footnotes.1",
         "Type.1",
         "Footnotes.2",
         "Type.2",
         "Footnotes.3",
         "Type.3",
         "Footnotes.4",
         "Type.4",
         "Footnotes.5", 
         "Type.5",
         "Footnotes.6", 
         "Type.6",
         "Footnotes.7", 
         "Type.7",
         "Footnotes.8",
         "Type.8",
         "Footnotes.9",
         "Type.9",
         "Footnotes.10",
         "Type.10",
         "Footnotes.11",
         "Type.11",
         "Footnotes.12",
         "Type.12",
         "Footnotes.13",
         "Type.13",
         "Footnotes.14",
         "Type.14",
         "Footnotes.15",
         "Type.15",
         "Footnotes.16",
         "Type.16",
         "Footnotes.17",
         "Type.17",
         "Footnotes.18",
         "Type.18",
         "Footnotes.19",
         "Type.19",
         "Footnotes.20",
         "Type.20",
         "Footnotes.21",
         "Type.21",
         "Footnotes.22",
         "Type.22",
         "Footnotes.23",
         "Type.23",
         "Footnotes.24",
         "Type.24",
         "Footnotes.25",
         "Type.25",
         "Footnotes.26",
         "Type.26",
         "Footnotes.27",
         "Type.27",
         "Footnotes.28",
         "Type.28",
         "Footnotes.29",
         "Type.29",
         "1991",
         "1992",
         "1993",
         "1994",
         "1995",
         "1996",
         "1997",
         "1998",
         "1999",
         #"2000",
         "2001",
         "2002",
         "2003",
         "2004",
         #"2005",
         "2006",
         "2007",
         "2008",
         "2009",
        # "2010"
         "2012",
         "2013",
         "2014",
         "2015",
         "2016",
         "2017",
         "2018",
         "2019"],inplace = True, axis = 1)
df1


CountryCode      Country  SeriesCode MDG  \
0              4  Afghanistan         567   Y   
1              4  Afghanistan         749   Y   
2              4  Afghanistan         616   Y   
3              8      Albania         567   Y   
4              8      Albania         749   Y   
..           ...          ...         ...  ..   
685          894       Zambia         749   Y   
686          894       Zambia         616   Y   
687          716     Zimbabwe         567   Y   
688          716     Zimbabwe         749   Y   
689          716     Zimbabwe         616   Y   

                                                Series     1990 Type     2000  \
0    Proportion of land area covered by forest, per...      2.1    E      2.1   
1    Carbon dioxide emissions (CO2), thousand metri...   2676.9    G    781.1   
2    Terrestrial and marine areas protected to tota...     0.36    C     0.36   
3    Proportion of land area covered by forest, per...     28.8    C     28.1   
4    Carbon dioxide emissions (CO2), thousand metri...     7488    G   3021.6   
..                                                 ...      ...  ...      ...   
685  Carbon dioxide emissions (CO2), thousand metri...   2445.9    G   1822.5   
686  Terrestrial and marine areas protected to tota...    36.05    C    36.06   
687  Proportion of land area covered by forest, per...     57.3    C     48.8   
688  Carbon dioxide emissions (CO2), thousand metri...  15504.1    G  13886.9   
689  Terrestrial and marine areas protected to tota...     16.9    C     16.9   

       2005    2010     2011  
0       2.1     2.1      NaN  
1    1338.5  8470.8  12251.4  
2       NaN     NaN      NaN  
3      28.5    28.3      NaN  
4    4253.7  4415.1   4668.1  
..      ...     ...      ...  
685  2277.2  2673.2   3047.3  
686     NaN     NaN      NaN  
687    44.6    40.4      NaN  
688   10715  9028.2   9860.6  
689     NaN     NaN      NaN  

[690 rows x 11 columns]

In [11]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CountryCode  690 non-null    int64 
 1   Country      690 non-null    object
 2   SeriesCode   690 non-null    int64 
 3   MDG          690 non-null    object
 4   Series       690 non-null    object
 5   1990         627 non-null    object
 6   Type         628 non-null    object
 7   2000         656 non-null    object
 8   2005         436 non-null    object
 9   2010         438 non-null    object
 10  2011         215 non-null    object
dtypes: int64(2), object(9)
memory usage: 59.4+ KB


In [12]:
#Checking unique values

#for i in df.columns:
   #print("{} is unique: {}".format(i, df[i].is_unique))
    df1.nunique( axis = 0, dropna = True)

CountryCode    230
Country        230
SeriesCode       3
MDG              1
Series           3
1990           515
Type             3
2000           568
2005           390
2010           386
2011           207
dtype: int64

In [14]:
#Statistical summary of my data

def new_data(df1): 
    data_df = {}
    data_df["null_sum"] = df1.isnull().sum()
    data_df["null_pct"] = df1.isnull().mean()
    data_df["dtypes"] = df1.dtypes
    data_df["count"] = df1.count()
    data_df['mean'] = df1.mean()
    data_df['median'] = df1.median()
    data_df['min'] = df1.min()
    data_df['max'] = df1.max()
    
    
    return pd.DataFrame(data_df)
new_data(df1)

null_sum  null_pct  dtypes  count        mean   median  \
1990               63  0.091304  object    627         NaN    22.40   
2000               34  0.049275  object    656         NaN    29.50   
2005              254  0.368116  object    436         NaN    73.00   
2010              252  0.365217  object    438         NaN    72.55   
2011              475  0.688406  object    215         NaN  7543.00   
Country             0  0.000000  object    690         NaN      NaN   
CountryCode         0  0.000000   int64    690  439.465217   439.00   
MDG                 0  0.000000  object    690         NaN      NaN   
Series              0  0.000000  object    690         NaN      NaN   
SeriesCode          0  0.000000   int64    690  644.000000   616.00   
Type               62  0.089855  object    628         NaN      NaN   

                                                           min  \
1990                                                       NaN   
2000                                                       NaN   
2005                                                       NaN   
2010                                                       NaN   
2011                                                       NaN   
Country                                            Afghanistan   
CountryCode                                                  4   
MDG                                                          Y   
Series       Carbon dioxide emissions (CO2), thousand metri...   
SeriesCode                                                 567   
Type                                                       NaN   

                                                           max  
1990                                                       NaN  
2000                                                       NaN  
2005                                                       NaN  
2010                                                       NaN  
2011                                                       NaN  
Country                                               Zimbabwe  
CountryCode                                                894  
MDG                                                          Y  
Series       Terrestrial and marine areas protected to tota...  
SeriesCode                                                 749  
Type                                                       NaN

In [17]:
new_data.columns = df1.columns[df1.isnull().any()]
df[new_data.columns].isnull().sum().sort_values(ascending = False).head(50)

2010    1
2005    1
2000    1
1990    1
2011    0
Type    0
dtype: int64

In [21]:
# Dropping columns with missing values
df2 = df1.dropna(how = 'all', axis = "columns")
df2.head()

CountryCode      Country  SeriesCode MDG  \
0            4  Afghanistan         567   Y   
1            4  Afghanistan         749   Y   
2            4  Afghanistan         616   Y   
3            8      Albania         567   Y   
4            8      Albania         749   Y   

                                              Series    1990 Type    2000  \
0  Proportion of land area covered by forest, per...     2.1    E     2.1   
1  Carbon dioxide emissions (CO2), thousand metri...  2676.9    G   781.1   
2  Terrestrial and marine areas protected to tota...    0.36    C    0.36   
3  Proportion of land area covered by forest, per...    28.8    C    28.1   
4  Carbon dioxide emissions (CO2), thousand metri...    7488    G  3021.6   

     2005    2010     2011  
0     2.1     2.1      NaN  
1  1338.5  8470.8  12251.4  
2     NaN     NaN      NaN  
3    28.5    28.3      NaN  
4  4253.7  4415.1   4668.1

In [23]:
#Checking number of values in each data type

df2.dtypes.value_counts()

object    9
int64     2
dtype: int64

In [24]:
#Checking all the object columns

df2.select_dtypes(include = ["object"]).columns

Index(['Country', 'MDG', 'Series', '1990', 'Type', '2000', '2005', '2010',
       '2011'],
      dtype='object')

In [26]:
#Checking correlation

df2.corr()

CountryCode    SeriesCode
CountryCode  1.000000e+00 -1.532248e-17
SeriesCode  -1.532248e-17  1.000000e+00

# 1.How many different countries are represented? How many missing values are there by country, year and series?

In [74]:
#Checking the number of countries represented
country_name = df2.Country.unique()

print("Number of countries in dataframes =", len(country_name))


Number of countries in dataframes = 230


In [89]:
#Checking missing values in the data frame
df2.isnull().sum().value_counts()

0      5
63     1
62     1
254    1
252    1
475    1
34     1
dtype: int64

There are 27 missing values in the data frame.

In [65]:
#Checking missing values in country

df2["Country"].isnull().any().sum()

0

There are no missing values in the Country column

In [64]:
#Checking missing values in series

df2["Series"].isnull().sum()

0

There are no missing values in the Series column.

In [63]:
#Checking missing values in 1990

df2["1990"].isnull().sum()

63

There are 63 missing values in 1990  

In [59]:
#Checking missing values in 2000

df2["2000"].isnull().value_counts()

False    656
True      34
Name: 2000, dtype: int64

There are 34 missing values in 2000

In [90]:
#Checking missing values in 2005

df2["2005"].isnull().value_counts()

False    436
True     254
Name: 2005, dtype: int64

There are 254 missing values in 2005

In [91]:
#Checking missing values in 2010

df2["2010"].isnull().value_counts()

False    438
True     252
Name: 2010, dtype: int64

There are 254 missing values in 2010

In [92]:
#Checking missing values in 2011

df2["2011"].isnull().value_counts()

True     475
False    215
Name: 2011, dtype: int64

There are 475 missing values in 2011

All in all year has 1078 missing values.

# 2.Who are the top and bottom 5 countries in terms of C02 emissions in 1990 and what are their emissions? How has this changed by 2011?
Do you spot any problems with the data? If so, fix it.

In [108]:
#Creating a new data frame for Country and 1990
df3 = df2.loc[:, ("Country","1990") ]

#Filling m,issing values with 1
df3["1990"] = df3["1990"].fillna(1)
df3

Country     1990
0    Afghanistan      2.1
1    Afghanistan   2676.9
2    Afghanistan     0.36
3        Albania     28.8
4        Albania     7488
..           ...      ...
685       Zambia   2445.9
686       Zambia    36.05
687     Zimbabwe     57.3
688     Zimbabwe  15504.1
689     Zimbabwe     16.9

[690 rows x 2 columns]

In [109]:
#Creating a new data frame for country and year 2011
df4 = df2.loc[:, ("Country","2011") ]

#Filling missing values using 1
df4["2011"] = df4["2011"].fillna(1)
df4

Country     2011
0    Afghanistan        1
1    Afghanistan  12251.4
2    Afghanistan        1
3        Albania        1
4        Albania   4668.1
..           ...      ...
685       Zambia   3047.3
686       Zambia        1
687     Zimbabwe        1
688     Zimbabwe   9860.6
689     Zimbabwe        1

[690 rows x 2 columns]